### PR TITLE
Storage hashed location refactoring

### PR DESCRIPTION
**Pull request by ehildenb**
This PR:

- Introduces basic accessor syntax in `hashed-locations.md`, this is the generic syntax for accessing member fields or struct members, as well as indexing into things like mapping or arrays.
- Changes the K code generation process to just generate the needed field names for the above generic accessor functionality.

At the moment, we lose the ability to generate the rules `rule Contract . field [ IDX ] => hashedLocation(...)`, which is needed for the `tests/specs/examples` proofs. So that needs to be added before moving forward with this PR.

### CODE REVIEW

**Comment by RaoulSchaffranek** [comment]
I get the following error when I run the end-to-end test:

~~~
> make tests/foundry/foundry.k.prove`
rm -rf tests/foundry/out
cd tests/foundry/ && forge build
[⠊] Compiling...
[⠢] Compiling 10 files with 0.8.15
[⠰] Solc 0.8.15 finished in 1.06s
Compiler run successful
make -C ./kevm_pyk venv-prod
make[1]: Entering directory '/home/raoul/rv/evm-semantics/kevm_pyk'
. /home/raoul/rv/evm-semantics/kevm_pyk/venv-prod/bin/activate
make[1]: Leaving directory '/home/raoul/rv/evm-semantics/kevm_pyk'
. ./kevm_pyk/venv-prod/bin/activate && kevm foundry-to-k tests/foundry/out --verbose --definition .build/usr/lib/kevm/haskell \
     --require lemmas/int-simplification.k --module-import INT-SIMPLIFICATION      \
     > tests/foundry/foundry.k
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/raoul/rv/evm-semantics/kevm_pyk/src/kevm_pyk/__main__.py", line 12, in <module>
    from .solc_to_k import Contract, contract_to_k, solc_compile
  File "/home/raoul/rv/evm-semantics/kevm_pyk/src/kevm_pyk/solc_to_k.py", line 32, in <module>
    from pyk.kastManip import abstract_term_safely, build_claim, substitute
ImportError: cannot import name 'build_claim' from 'pyk.kastManip' (/home/raoul/rv/evm-semantics/kevm_pyk/venv-prod/lib/python3.8/site-packages/pyk/kastManip.py)
make: *** [Makefile:445: tests/foundry/foundry.k] Error 1
~~~

Also, when I get an error when I try to `make` pyk:

~~~
> cd kevm_pyk
> make
. /home/raoul/rv/evm-semantics/kevm_pyk/venv-dev/bin/activate && isort --check src
. /home/raoul/rv/evm-semantics/kevm_pyk/venv-dev/bin/activate && flake8 src
. /home/raoul/rv/evm-semantics/kevm_pyk/venv-dev/bin/activate && mypy src
src/kevm_pyk/solc_to_k.py:32: error: Module "pyk.kastManip" has no attribute "build_claim"
Found 1 error in 1 file (checked 5 source files)
make: *** [Makefile:56: check-mypy] Error 1
~~~

**Comment by ehildenb** [comment]
@RaoulSchaffranek you need to rebuild your pyk library, there was a KEVM submodule update.

Start by cleaning:

```
git clean -dffx -- kevm_pyk/ tests/
```

Then rebuild:

```
make -C ./kevm_pyk venv-dev venv-prod
```

**Comment by RaoulSchaffranek** [comment]
I made some progress, and I hope I now have a clean environment. For reference, here are the commands I ran:

~~~bash
git submodule update --init --recursive
git reset --hard
git clean -dffx -- kevm_pyk/ tests/
make clean
make k-deps
make build-haskell
make -C ./kevm_pyk
make -C ./kevm_pyk venv-dev venv-prod
make build-kevm
~~~

But I still get an error when running the end-to-end test:

~~~
> make tests/foundry/foundry.k.prove
rm -rf tests/foundry/out
cd tests/foundry/ && forge build
[⠊] Compiling...
[⠢] Compiling 2 files with 0.8.15
[⠆] Solc 0.8.15 finished in 17.75ms
Compiler run successful
make -C ./kevm_pyk venv-prod
make[1]: Entering directory '/home/raoul/rv/evm-semantics/kevm_pyk'
. /home/raoul/rv/evm-semantics/kevm_pyk/venv-prod/bin/activate
make[1]: Leaving directory '/home/raoul/rv/evm-semantics/kevm_pyk'
. ./kevm_pyk/venv-prod/bin/activate && kevm foundry-to-k tests/foundry/out --verbose --definition .build/usr/lib/kevm/haskell \
     --require lemmas/int-simplification.k --module-import INT-SIMPLIFICATION      \
     > tests/foundry/foundry.k
INFO 2022-08-05 10:41:02,698 __main__ - Processing contract file: tests/foundry/out/Token.sol/Token.json
INFO 2022-08-05 10:41:02,699 __main__ - Produced contract module: TOKEN-BIN-RUNTIME
INFO 2022-08-05 10:41:02,699 __main__ - Processing contract file: tests/foundry/out/Token.t.sol/TokenTest.json
INFO 2022-08-05 10:41:02,717 __main__ - Produced contract module: TOKENTEST-BIN-RUNTIME
. ./kevm_pyk/venv-prod/bin/activate && kevm kompile tests/foundry/foundry.k --pyk --backend haskell --definition tests/foundry/kompiled \
    --main-module VERIFICATION --syntax-module VERIFICATION               \
    --concrete-rules-file tests/foundry/concrete-rules.txt                \
    -I /usr/lib/kevm/include/kframework -I /usr/lib/kevm/blockchain-k-plugin/include/kframework
. ./kevm_pyk/venv-prod/bin/activate && kevm prove tests/foundry/foundry.k --pyk --backend haskell --definition tests/foundry/kompiled \
     --spec-module TOKENTEST-BIN-RUNTIME-SPEC
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/raoul/rv/evm-semantics/kevm_pyk/src/kevm_pyk/__main__.py", line 158, in <module>
    main()
  File "/home/raoul/rv/evm-semantics/kevm_pyk/src/kevm_pyk/__main__.py", line 102, in main
    final_state = kevm.prove(spec_file, spec_module_name=spec_module, args=prove_args, haskell_args=haskell_args, rule_profile=spec_file.with_suffix('.rule-profile'))
  File "/home/raoul/rv/evm-semantics/kevm_pyk/venv-prod/lib/python3.8/site-packages/pyk/ktool/kprove.py", line 164, in prove
    productivity = total_success_time / (total_success_time + total_failure_time)
ZeroDivisionError: float division by zero
make: *** [Makefile:453: tests/foundry/foundry.k.prove] Error 1
~~~

**Comment by ehildenb** [comment]
`tests/foundry/foundry.k.prove` is disabled as a test now.

You need to do `make tests/foundry/foundry.k.check` and `tests/foundry/kompiled/timestamp` to check the functionality here.

**Comment by ehildenb** [comment]
Sorry, accidentally pressed close.

File: kevm_pyk/src/kevm_pyk/solc_to_k.py [hunk 1]
```diff
@@ -219,11 +167,17 @@ def method_sentences(self) -> List[KSentence]:
...
     @property
     def field_sentences(self) -> List[KSentence]:
-        field_access_production: KSentence = KProduction(KSort('Int'), [KNonTerminal(self.sort), KTerminal('.'), KNonTerminal(self.sort_field)], klabel=self.klabel_field, att=KAtt({'macro': ''}))
-        res: List[KSentence] = [field_access_production]
-        res.extend(prod for field in self.fields for prod in field.productions)
-        res.extend(rule for field in self.fields for rule in field.rules(KApply(self.klabel), self.klabel_field))
-        return res if len(res) > 1 else []
+        prods: List[KSentence] = [self.subsort_field]
+        rules: List[KSentence] = []
+        for field, offset in self.fields.items():
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply kevm_pyk/src/kevm_pyk/solc_to_k.py:172]: Yes, I can do that, but if a renaming is the only change you want, you shoudl approve and put a message "Approving assuming that you will make these changes." That's how we do it with all our other repos, and leads ot much faster development.

[the remark this answers is not in the document]
Comment (ehildenb) [reply kevm_pyk/src/kevm_pyk/solc_to_k.py]: Done.

**Review by RaoulSchaffranek — APPROVED** [review]
Alright, got it. Refactoring looks good to me!

**Pull request merged**